ethers-solc/src/resolver.rs

```diff
@@ -0,0 +1,518 @@
+                let dep = *dep;
+                if dep >= num_input_files {
+                    // library import
+                    if let Some(node) = all_nodes.remove(&dep) {
+                        sources.insert(node.path, node.source);
+                        insert_imports(dep, all_nodes, sources, edges, num_input_files);
+                    }
+                }
+            }
+        }
+
+        let versioned_nodes = self.get_input_node_versions(offline)?;
+        let Self { nodes, edges, num_input_files, .. } = self;
+        let mut versioned_sources = HashMap::with_capacity(versioned_nodes.len());
+        let mut all_nodes = nodes.into_iter().enumerate().collect::<HashMap<_, _>>();
+        for (version, input_node_indices) in versioned_nodes {
+            let mut sources = Sources::new();
+            for idx in input_node_indices {
+                // insert the input node
+                let node = all_nodes.remove(&idx).unwrap();
+                sources.insert(node.path, node.source);
+                insert_imports(idx, &mut all_nodes, &mut sources, &edges, num_input_files);
+            }
+            versioned_sources.insert(version, sources);
+        }
+        Ok(VersionedSources { inner: versioned_sources, offline })
+    }
+
+    fn format_imports_list<W: std::fmt::Write>(
+        &self,
+        idx: usize,
+        f: &mut W,
+    ) -> std::result::Result<(), std::fmt::Error> {
+        let node = self.node(idx);
+        for dep in self.imported_nodes(idx) {
+            let dep = self.node(*dep);
+            writeln!(
+                f,
+                "  {} ({:?}) imports {} ({:?})",
+                utils::source_name(&node.path, &self.root).display(),
+                node.data.version,
+                utils::source_name(&dep.path, &self.root).display(),
+                dep.data.version
+            )?;
+        }
+        for dep in self.imported_nodes(idx) {
+            self.format_imports_list(*dep, f)?;
+        }
+
+        Ok(())
+    }
+
+    /// Filters incompatible versions from the `candidates`.
+    fn retain_compatible_versions(
+        &self,
+        idx: usize,
+        candidates: &mut Vec<&crate::SolcVersion>,
+        traversed: &mut std::collections::HashSet<(usize, usize)>,
+    ) -> std::result::Result<(), String> {
+        let node = self.node(idx);
+
+        if let Some(ref req) = node.data.version_req {
+            candidates.retain(|v| req.matches(v.as_ref()));
+        }
+        for dep in self.imported_nodes(idx) {
+            let dep = *dep;
+            // check for circular deps which would result in endless recursion SO here
+            // a circular dependency exists, if there was already a `dependency imports current
+            // node` relationship in the traversed path
+            if traversed.contains(&(dep, idx)) {
+                let mut msg = String::new();
+                self.format_imports_list(dep, &mut msg).unwrap();
```

Comment:
The main reason for `format_imports_list` is for pretty error printing? Let's maybe add that as a note?